### PR TITLE
[Merged by Bors] - feat(ModelTheory/Equivalence): Implication of formulas modulo a theory

### PR DESCRIPTION
**Pull request by awainverse**
Defines `FirstOrder.Language.Theory.Implies`: `φ ⟹[T] ψ` indicates that `φ` implies `ψ` in models of `T`.
Relates `Theory.Implies` to `Theory.SemanticallyEquivalent`.
Proves lemmas that will set up a `BooleanAlgebra` instance on formulas modulo a theory

---

- [x] depends on: #16799 

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/ModelTheory/Equivalence.lean
Comment (awainverse) [reply Mathlib/ModelTheory/Equivalence.lean:77]: I'm not sure which iff you're looking for, sorry - what lemma are you looking for?

File: Mathlib/ModelTheory/Equivalence.lean
Comment (awainverse) [reply Mathlib/ModelTheory/Equivalence.lean:118]: I've just tried a range of precedences, and nothing seemed to work. Do you have a suggestion of something to change other than the number `25`?

File: Mathlib/ModelTheory/Equivalence.lean
Comment (YaelDillies) [reply Mathlib/ModelTheory/Equivalence.lean:75]: `(a sup b -> c) <-> (a -> c) and (b -> c)`

File: Mathlib/ModelTheory/Equivalence.lean
Comment (YaelDillies) [reply Mathlib/ModelTheory/Equivalence.lean:116]: I'll try playing with precedences

File: Mathlib/ModelTheory/Equivalence.lean
Comment (awainverse) [reply Mathlib/ModelTheory/Equivalence.lean:75]: Oh of course - I somehow convinced myself that you were looking for a `SemanticallyEquivalent` statement about `sup`.

Comment (awainverse) [reply Mathlib/ModelTheory/Equivalence.lean:75]: (I also added `implies_inf_iff`.)

**Comment by YaelDillies** [comment]
I don't quite like how the following square is looking:
`Implies`/`Imp` | `Language.Theory.Implies`
`Iff` | `Language.Theory.SemanticallyEquivalent`

What do you think of renaming the right column so that it looks more consistent? Here are a few suggestions (all in the `Language.Theory` namespace):
* `SemanticallyImplies`, `SemanticallyEquivalent`
* `SemanticImp`, `SemanticIff`
* `Imp`, `Iff`

**Comment by awainverse** [comment]
I agree that making these match makes sense - I think it should be down to the last two options, so the question is just whether to specify `Semantic`. The reason for doing this is because I hope that at some point, it would be good to have a first-order proof system, and in that case, there's a distinction between a theory modelling an implication/equivalence and the theory proving it. These turn out to be the same, of course, because of the completeness theorem, so it's just a question of which should be the _definition_ of `Theory.Imp` and `Theory.Iff`. At the moment, I'd say I over-futureproofed `SemanticallyEquivalent`, so probably this should just be `Imp` and `Iff`.

**Comment by YaelDillies** [comment]
I'll let you make the final call. It's easy enough to rename the predicate later (and probably we won't need deprecations since you are currently the sole downstream user of this material)

**Comment by awainverse** [comment]
I've finally gone and used the simplest possible choice on the names. Hopefully I've capitalized (or uncapitalized) the names of lemmas appropriately.

File: Mathlib/ModelTheory/Complexity.lean
Comment (YaelDillies) [inline Mathlib/ModelTheory/Complexity.lean:308]:
```suggestion
theorem iff_toPrenex (φ : L.BoundedFormula α n) :
```

**Review by YaelDillies — COMMENTED** [review]
Technically, `Iff` already exists in the root namespace, so this should be protected. I think that should be fine this you only use that declaration through its notation?

**Review by YaelDillies — APPROVED** [review]
Let's go!

maintainer merge

File: Mathlib/ModelTheory/Equivalence.lean
Comment (b-mehta) [inline Mathlib/ModelTheory/Equivalence.lean:66]: I'd prefer to name this `sup_imp_left` in analogy with `sup_le_left` for orders

Comment (b-mehta) [reply Mathlib/ModelTheory/Equivalence.lean:66]: (and similar below)

File: Mathlib/ModelTheory/Equivalence.lean
Comment (b-mehta) [inline Mathlib/ModelTheory/Equivalence.lean:112]:
```suggestion
    ⊤ ⟹[T] φ ⊔ ∼φ := fun M v xs => by
  simp only [BoundedFormula.realize_imp,
    BoundedFormula.realize_top, BoundedFormula.realize_sup, BoundedFormula.realize_not,
    true_implies]
  exact or_not
```

File: Mathlib/ModelTheory/Equivalence.lean
Comment (b-mehta) [inline Mathlib/ModelTheory/Equivalence.lean:105]:
```suggestion
    φ ⊓ ∼φ ⟹[T] ⊥ := fun M v xs => by 
  simp only [BoundedFormula.realize_imp,
    BoundedFormula.realize_inf, BoundedFormula.realize_not, and_not_self,
    BoundedFormula.realize_bot, imp_self]
```

File: Mathlib/ModelTheory/Equivalence.lean
Comment (b-mehta) [inline Mathlib/ModelTheory/Equivalence.lean:102]: This one and the next are `iff` in general, right? It seems more sensible to just give that version, then this one should be easily derivable from it

File: Mathlib/ModelTheory/Equivalence.lean
Comment (YaelDillies) [reply Mathlib/ModelTheory/Equivalence.lean:66]: You mean [`le_sup_left](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Order/Lattice.html#le_sup_left)`?

Comment (b-mehta) [reply Mathlib/ModelTheory/Equivalence.lean:66]: oops lol

File: Mathlib/ModelTheory/Equivalence.lean
Comment (awainverse) [reply Mathlib/ModelTheory/Equivalence.lean]: I've now made both of these `iff`s and adjusted #16801 accordingly.

**Comment by YaelDillies** [comment]
maintainer merge

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**
